src/main/java/org/tugraz/sysds/runtime/data/LibTensorAgg.java

```diff
@@ -0,0 +1,396 @@
+		}
+
+		//core aggregation
+		boolean lastRowCorr = (aop.correctionLocation == PartialAggregate.CorrectionLocationType.LASTROW);
+		boolean lastColCorr = (aop.correctionLocation == PartialAggregate.CorrectionLocationType.LASTCOLUMN);
+		if( !in.isSparse() && lastRowCorr )
+			aggregateBinaryTensorLastRowDenseGeneric(in, aggVal);
+		else if( in.isSparse() && lastRowCorr )
+			throw new DMLRuntimeException("Aggregation on sparse tensors not yet supported.");
+		else if( !in.isSparse() && lastColCorr )
+			aggregateBinaryTensorLastColDenseGeneric(in, aggVal);
+		else if( in.isSparse() && lastColCorr )
+			throw new DMLRuntimeException("Aggregation on sparse tensors not yet supported.");
+	}
+
+	/**
+	 * Get the aggregation type from the unary operator.
+	 *
+	 * @param op the unary operator
+	 * @return the aggregation type
+	 */
+	private static AggType getAggType(AggregateUnaryOperator op) {
+		ValueFunction vfn = op.aggOp.increOp.fn;
+
+		//(kahan) sum
+		if (vfn instanceof KahanPlus)
+			return AggType.KAHAN_SUM;
+		return AggType.INVALID;
+	}
+
+	/**
+	 * Determines whether the unary operator is supported.
+	 * @param op the unary operator to check
+	 * @return true if the operator is supported, false otherwise
+	 */
+	public static boolean isSupportedUnaryAggregateOperator( AggregateUnaryOperator op ) {
+		AggType type = getAggType( op );
+		return (type != AggType.INVALID);
+	}
+
+	/**
+	 * Aggregate a subset of rows of a dense tensor block.
+	 * @param in the tensor block to aggregate
+	 * @param out the aggregation result with correction
+	 * @param aggtype the type of aggregation to use
+	 * @param fn the function to use
+	 * @param ixFn the IndexFunction to use
+	 * @param rl the lower index of rows to use
+	 * @param ru the upper index of rows to use (exclusive)
+	 */
+	private static void aggregateUnaryTensorDense(TensorBlock in, TensorBlock out, AggType aggtype, ValueFunction fn,
+	                                              IndexFunction ixFn, int rl, int ru) {
+		//note: due to corrections, even the output might be a large dense block
+		DenseBlock a = in.getDenseBlock();
+		DenseBlock c = out.getDenseBlock();
+		if (aggtype == AggType.KAHAN_SUM) {
+			KahanObject kbuff = new KahanObject(0, 0);
+			if (ixFn instanceof ReduceAll) // SUM
+				d_uakp(a, c, kbuff, (KahanPlus) fn, rl, ru);
+		}
+		// TODO other aggregations
+	}
+
+	/**
+	 * Add two tensor-blocks, which contain the result of an aggregation with correction in the last row, together.
+	 * @param in the tensor block to add
+	 * @param aggVal the tensor block to which the first should be added
+	 */
+	private static void aggregateBinaryTensorLastRowDenseGeneric(TensorBlock in, TensorBlock aggVal) {
+		if( in._denseBlock==null || in.isEmpty(false) )
+			return;
+
+		final int m = in.getDim(0);
+		final int n = in.getDim(1);
+		final int cix = (m-1)*n;
+
+		double[] a = in.getDenseBlock().valuesAt(0);
```

Comment:
Please add a TODO for generalize for large blocks if you use it for local operations of large matrices (not vectors).